### PR TITLE
Light refactoring of implicit arguments

### PR DESCRIPTION
**Pull request by jashug**
**Kind:** refactoring

These changes should be invisible from outside the `Impargs` module.
The general idea is to split computation of implicit arguments into several (completely independent) passes (is it rigid, what are the names, are they inferable),
and split `set_manual_implicits` into first mapping `ExplByPos` and `ExplByName` to a list of requested implicit arguments (in `flatten_explicitations`), and then performing the merge (which has really weird and inconsistent logic).
We also remove a few unused parameters.

### CODE REVIEW

**Comment by herbelin** [comment]
If the changes are invisible from outside (or even if there had been changes properly motivated), this is good to me, and thanks for the cleaning. Don't know the opinion of @mattam82, who also contributed to `impargs.ml`, though.

File: interp/impargs.ml [hunk 1]
```diff
@@ -213,6 +213,8 @@ let add_free_rels_until strict strongly_strict revpat bound env m pos acc =
   let () = if not (Vars.noccur_between 1 bound m) then frec true (env,1) m in
   acc
 
+(* calculate the list of implicit arguments *)
```
Comment (SkySkimmer) [inline interp/impargs.ml]: "compute" is more idiomatic english.

File: interp/impargs.ml [hunk 2]
```diff
@@ -225,7 +227,18 @@ let rec is_rigid_head t = match kind t with
...
 
-(* calcule la liste des arguments implicites *)
+let is_rigid env t =
+  let open Context.Rel.Declaration in
+  let rec aux env t =
+    let t = whd_all env t in
+    match kind t with
+    | Prod (na,a,b) -> aux (push_rel (LocalAssum (na,a)) env) b
+    | _ -> is_rigid_head t
+  in
```
Comment (SkySkimmer) [inline interp/impargs.ml]: You could use `Reductionops.splay_prod` instead of rolling a custom version.

**Comment by jashug** [comment]
Rebased after `impargs.ml` was econstrized by ejgallego in
https://github.com/coq/coq/commit/886a9c2fb25e32bd87b3fce38023b3e701134d23#diff-1e0fbe8a87db8a25a4a7a500ff1ad50e

@maximedenes When you have time, I'd like to hear your thoughts.

**Comment by herbelin** [comment]
Does someone want to be "assignee"? I can do it otherwise. Status: pending at least from a review by @maximedenes.

**Comment by herbelin** [comment]
@maximedenes: are you still willing to do a review on this PR?

**Comment by maximedenes** [comment]
It was on my screen when you typed your last message :) So yes.

**Review by maximedenes — APPROVED** [review]
My understanding of this code is limited, but this PR looks like a good first cleaning step. In particular, separating the passes is very good. Don't hesitate to go further in subsequent PRs, even if simplifying the logic means changing a bit the functionality

**Pull request merged**